src/Symfony/Component/Security/Core/Authorization/Voter/CallableVoter.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Authorization\Voter;
+
+use Symfony\Component\Security\Core\Authentication\AuthenticationTrustResolverInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Security\Core\Authorization\IsGrantedPayload;
+use Symfony\Component\Security\Core\Role\RoleHierarchyInterface;
+
+/**
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ */
+final class CallableVoter implements CacheableVoterInterface
+{
+    public function __construct(
+        private AuthenticationTrustResolverInterface $trustResolver,
+        private AuthorizationCheckerInterface $authChecker,
+        private ?RoleHierarchyInterface $roleHierarchy = null,
+    ) {
+    }
+
+    public function supportsAttribute(string $attribute): bool
+    {
+        return false;
+    }
+
+    public function supportsType(string $subjectType): bool
+    {
+        return true;
+    }
+
+    public function vote(TokenInterface $token, mixed $subject, array $attributes): int
+    {
+        $result = VoterInterface::ACCESS_ABSTAIN;
+        $payload = null;
+        foreach ($attributes as $attribute) {
+            if (!\is_callable($attribute)) {
+                continue;
+            }
+
+            $payload ??= $this->getPayload($token, $subject);
+
+            $result = VoterInterface::ACCESS_DENIED;
+            if ($attribute($payload)) {
```

Comment:
I'm not happy with that payload class either and would prefer passing multiple arguments instead.